drivers/soundwire/intel.c

```diff
@@ -876,21 +932,28 @@ static int intel_trigger(struct snd_pcm_substream *substream, int cmd,
 	struct sdw_cdns_dma_data *dma;
 	int ret;
 
+	dev_err(dai->dev, "%s: %s: start\n", __func__, dai->name);
+
 	dma = snd_soc_dai_get_dma_data(dai, substream);
 	if (!dma) {
 		dev_err(dai->dev, "failed to get dma data in %s", __func__);
 		return -EIO;
 	}
 
 	switch (cmd) {
+	case SNDRV_PCM_TRIGGER_RESUME:
+		dev_err(dai->dev, "%s: %s: resume\n", __func__, dai->name);
+		/* fallthrough */
 	case SNDRV_PCM_TRIGGER_START:
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
-	case SNDRV_PCM_TRIGGER_RESUME:
 		ret = sdw_enable_stream(dma->stream);
 		break;
 
-	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
 	case SNDRV_PCM_TRIGGER_SUSPEND:
+		dev_err(dai->dev, "%s: %s: suspend\n", __func__, dai->name);
+		dma->suspend_reinitialize = true;
```

Comment:
@plbossart set suspend_reinitialize = true here may not be the best place. The reason we need to reinitialize cdns is when system is suspended. So I think it is better to set suspend_reinitialize = true in intel_suspend(). We may need https://github.com/bardliao/linux/commit/d0247d3e90da8174f5d0fc6a00e081fb911ced7b

Reply:
No, I don't think we can set suspend_reinitialize in intel_suspend. The reason is that we need to reprogram the ALH stream ID, and if you have more than one DAI for a given device  then with your patch you would only restore the first ALH stream ID. It's got to be done on a DAI-basis, so the flag has to be stored in the dma structure.

Reply:
That said, you are right @bardliao that if the DAI is prepared but not triggered, it still needs to be re-initialized. I guess the solution is a mix of yours and mine, we can't use TRIGGER suspend but can't use the .suspend either since each DAI needs to be reinitialized.